### DISCUSSION TITLE
Custom domains per collection

### DISCUSSION THREAD

**Discussion by quickstar** [outline/outline#4077] [Ideas]
Hi everyone

We did start using outline primarily to write internal technical documentation.
We're now thinking about writing project/product documentation in outline as well.

Especially for product documentation we'd love to be able to share a whole collection publicly.
Further more it would be amazing if we share a collection publicly to have the possibility to add custom domains per collection.

For example:
Outline instance: https://outline.mydomain.org
Collections:
- Awesome-Product
  - https://outline.mydomain.org/collection/awesomeproduct-coIve9wob2 (internal editor link)
  - https://doc.awesomeproduct.org (publicly available readonly link)

I think this would be a unique and very handy feature. What do you think about it?

**Comment by tommoor** [comment]
See outstanding issue: https://github.com/outline/outline/issues/1160

**Reply by quickstar** [reply]
> See outstanding issue: #1160

Jep almost exactly what I was looking for, tnx! I commented on this case regarding custom domains! this discussion can be closed, I guess :-)

**Comment by almereyda** [comment]
Custom domains comes with one caveat for self-hosters: It is not possible to know the custom domains beforehand, which comes with implications for certificate generation.

- Self-hosters on Docker most likely configure their web proxy with labels, which are configured ahead of runtime.
- Kubernetes users can make use of the Kubernetes API to generate expected certificates with cert-manager and deploy them into a custom ingress rule.

One way that seems possible is to allow custom *sub*-domains of a preconfigured outline subdomain. E.g.:

- Outline runs on `docs.example.net`
- Public collections are published on subdomains, like `*.docs.example.net`, `*.docs.example.com` or `*.pages.example.net`. Using a separate base domain than Outline itself for publishing user-generated content is generally considered a good convention, in order to use a different security context in the browser and to avoid potential XSS attacks, e.g. stealing cookies.

When using a subdomain wildcard domain, it is possible to generate wildcard certificates, by using the DNS-01 challenge.

Alternatively the GitLab Pages path could be used as inspiration:

- Check DNS settings ahead of deploying web routes and generating certificates, eventually involving a custom TXT record to validate ownership of the (sub-)domain towards Outline
- Use a separate process with its own IP that handles serving web traffic, TLS termination and certificate generation on its own.

A custom subdomain under a specified domain with using a wildcard together with custom domains, incl. apex domains with on-the-fly certificates generated by a separate process, is possible with a separate SNI proxy, e.g. with Traefik on Docker:

```mermaid
flowchart TD
    Traffic -->|ALIAS pages.example.com| SNIproxy
    SNIproxy{"`Traefik TCP, **IP 1**`"} -->|HostSNI pages.example.com *.pages.example.com| HTTPSproxy("`Traefik HTTPS, **IP 2**`")
    HTTPSproxy -->|HostRegexp pages.example.com *.pages.example.com| Pages("`**Outline Pages daemon**
(1) serves autogenerated routes on *.pages.example.com
(2) after validation serves custom domains and custom certificates on`")
    SNIproxy -->|HostSNI *| Pages
```

I'd suggest to set up a GitLab Pages site with a custom domain to test their flow and what people might be expecting from such a feature.

---

*Moved from https://github.com/outline/outline/issues/1160#issuecomment-2106975044*

**Comment by almereyda** [comment]
Another interesting implementation that allows to implement this use case is using Caddy with its on demand TLS provisioning, eventually dependant on an `ask` services that confirms or denies requests for issuing certificates of validated domains.

- [bearblog/Caddyfile at master · HermanMartinus/bearblog](https://github.com/HermanMartinus/bearblog/blob/master/Caddyfile)
- [Automatic HTTPS — Caddy Documentation](https://caddyserver.com/docs/automatic-https#using-on-demand-tls)
- [Avoid specifying on demand 'ask' endpoint when using my own issuer · Issue #5986 · caddyserver/caddy](https://github.com/caddyserver/caddy/issues/5986)
- [caddytls: Make on-demand 'ask' permission modular by mholt · Pull Request #6055 · caddyserver/caddy](https://github.com/caddyserver/caddy/pull/6055)
- [How We Use Caddy to Provide Custom Domains for Our Clients](https://pirsch.io/blog/how-we-use-caddy-to-provide-custom-domains-for-our-clients/)

This could be a little boolean web service, endpoint or worker function, which acts on existing state, leaving the validation procedure decoupled.

**Comment by EduNetEurope** [comment]
Are there any updates regarding custom domains per collections, @tommoor?
